### DISCUSSION TITLE
Attribute "Change"/"Update" behaviour?

### DISCUSSION THREAD

**Discussion by Apollon77** [mfucci/node-matter#69] [General]
The current "Attribute implementation" in https://github.com/mfucci/node-matter/blob/main/src/matter/cluster/server/Attribute.ts#L24 only supports "changing settings" because if the value did not change the whole logic is not executed.

In smart home logics you usually differentiate between "change events" (aka "value changes") and "update events" (also "retriggering the same value" is  valid here). And logics react on that. best example is amovement sensor that goes to true and after some time to false when movement is over, BUT can also "retrigger" a true if the movement is still happening.
Such "update events" are relevant for certain logics because it is sometimes important to also get "the same value again".

Because of that I propose to remove the above mentioned line or make sure (if it is about "do we need to increase version or not) to have the handlers also called for the same value.

**Comment by mfucci** [comment]
Are you sure this is how this is supposed to work from Matter point of view?
I don't know what it is supposed to be but I implemented this was to simplify the code and to improve the performance.

From what I understood, Matter also has a notion of "events" (not implemented yet in node-matter) that should fulfill this use case.
Also, when the subscription logic is to send periodically send again the current value, even if it has not changed.

I am travelling for a week, so I cannot check with the official Matter sample device how it behaves (does turning on the on/off cluster when it is already on send back a update on the on attribute subscription? ). I can check that when I am back or let me know if you can check on your side.

**Comment by Apollon77** [comment]
In fact with chiptool I can set "on" and  then send "on" again and it is sent out ... :-) So si think it is not limited and depends on the device what it does. From a pure "general smart home devices perspective" I just know several use cases where also "triggering the same state" is important. E.g. when devices use a "turn off timer after turning it on" it partially means that this time is re-set when you tun on again while the turn-off timer is running. I think it is not that relevant for performance (at least not from my experience).

Good example maybe here from specs:

> 1.5.7.6. OnWithTimedOff Command
The OnWithTimedOff command allows devices to be turned on for a specific duration with a guarded off duration so that SHOULD the device be subsequently turned off, further OnWithTimed­ Off commands, received during this time, are prevented from turning the devices back on. Further OnWithTimedOff commands received while the server is turned on, will update the period that the device is turned on.

So the cases will become more complex and "retriggering" things - even if only a sub value changes might get interesting when we limit it that way.

For on we just have this info in specs:

> On receipt of the On command, a server SHALL set the OnOff attribute to TRUE.
Additionally, when the OnTime and OffWaitTime attributes are both supported, if the value of the OnTime attribute is equal to 0, the server SHALL set the OffWaitTime attribute to 0.

In fact the question for "Server" cases is not 100% about this, but also about what we ofer the users of the library that in fact implement the logic of the device. So it is a bit "over" the pure Natter specs... matter just say" when turning on you set the on attribute to TRUE" ... in fact the current code is "ignoring a "reset to TRUE"".

When I think about myself that plans to offer ioBroker devices via that lib I would need to have the events in all cases ...

**Comment by mfucci** [comment]
I am still not sure I understand exactly the need:

- right now the library doesn't trigger listeners when an attribute is set to its current value
- you are planing on using listeners on attributes to glue with the lower layer in ioBroker (Attribute.addListener on the server-side)

Am I correct?


To consider for using the library:

- the OnOff attribute is read-only for the client, so its value can only be modified on the server-side
- the On / Off / Toggle commands will be called each time to change the state, so the glue logic can be implemented there to hook with a lower layer
- to implement something listening to events, it would be probably better to actually use events. Attribute value is a state, so not chaging when set again to the same value...

From Matter point of view: I am still unsure what Matter expect in that case.
An attribute value has a version. Does the version increases when the attribute is set again to it current value? In the case of the OnOff cluster, does calling on() command on a light already one increases the OnOff attribute version?

Let's separate this into two separate points:

1 - what is the expected behavior for Matter? The current implementation might be incorrect
2 - what API are needed to be able to use the library: I can add another type of listeners on Attribute that will trigger regardless if the value changes or not

For 1, I cannot test right now since I am not home for a few more days...
For 2, let me know what additional API should be added on the library to fulfill your use case (or you can add it in a PR)

**Reply by Apollon77** [reply]
Yes basically you got it right.

So I also try to structure it a bit better:

### Cluster version
According to the specs it is right that when the value doesn't change the "cluster version" also should not change because in fact the overall state of the cluster do not change.

> A cluster data version is a metadata increment-only counter value, maintained for each cluster instance. A cluster data version represents an exact & coherent state of cluster attribute data at present. 
> A cluster data version is surfaced in the Interaction Model when data is requested. It is used to optimize data read transactions by reducing the need to send the same data. 
> A cluster data version SHALL increment or be set (wrap) to zero if incrementing would exceed its maximum value. A cluster data version SHALL be maintained for each cluster instance. A cluster data version SHALL be initialized randomly when it is first published. A cluster data version SHALL be incremented if any attribute data changes.

So yes the current check if the vaue is the same is correct for the increase of the data version.
(PS: We currently miss the overflow and random initialization logics)

I did read again a bit and in fact the Matter specifications just tell "set the attribute to value X". Some commands react differently if the value is a defined one. So in fact it is up to us to have that in the library how it makes sense on how to bring "state values to lower layers". We can also say that we leave it as now and see if we get issues ... but no idea if changing the behaviour later will then bring other incompatibilites for the users. Thats why I brought it up that early.
It is nothing where the Matter specs will really help us because this is now "implementation specific" :-)

### When using this Library as "Device"
This is one usecase which allows to offer devices available in ioBroker to be controlled via other Matter "Clients". Here my idea so far was that I bind listeners to the attributes and also set the "ioBroker state connected to the attribute" as soon as a control command comes in. In this case I see the mentioned issue with the current implementation that the listeners are not triggered when the same value is set again.

Because of the fact that the commands and their behaviour are defined in Matter specs I would not start binding the commands to lower layer and stay with "Just the attributes". Thats why I think the above option to have attribute value listeners makes a lot of sense.

So my proposal would be to always call the listeners (or emit events - but when using events the listeners should be replaced by events) on "setting a value" irrelevant if changed or not and let the lower level decide if same value is handled or ignored.

And we also need "the way back". When the device decides that it should be turned off now (e.g. because TimeOff was used and the time is over now) then the attribute will change triggered by library ocde or maybe also "lower layer code" and is sent out to all potential subscribers.

### When using this Library as "Controller"
The other usage I plan (once functionality available) is to use the library as basis for a "controller" approach to be able to control other devices after pairing. (I maintain a library and ioBroker adapter that does that with Homekit protocol already).

Here the idea is to iterate over the Cluster structure received for the paired device(s), build that structure in ioBroker and in fact also map for Attribute changes based on events/listeners or set Attributed when controlling :-)

Long story short: I think it makes sense to have listeners to be called irrelevant if value changes or not - or events for that. So my usecase would be fullfilled by adjusting that one handling to only increase version when value change but always trigger listener :-)

**Comment by mfucci** [comment]
Here is what I suggest.

When using this Library as "Device"
I add two kinds of listeners on attributes:

- subscribe<T>(listener: (value: T, version: number) => void): called only when the data version changes, it is used for the attribute subscription. meant to be used internally by the library.
- addListener(listener: (newValue: T, oldValue: T) => void): called each time that set() is called on the attribute regardless if the value is the same

To change the attribute value locally, set() can be called.

When using this Library as "Controller"
The cluster client callback is called each time the subscription sends a new value. Official Matter devices will only send an update when the value changes or periodically, so I think you will have to stick with this behavior...

Sounds good?

So I think the work is to rename the current API and add a new one.

**Reply by Apollon77** [reply]
Yes sounds good. Thank you

**Comment by mfucci** [comment]
Created a PR: https://github.com/mfucci/node-matter/pull/80
